File: algorithm_library/python/graph/cycle_detect.py
vis = [False]*(N+1)
fin = [False]*(N+1)
history = deque()

# 頂点 v の連結成分内の閉路を返す
def cycle_detect(v):
  sta = deque()
  sta.append((v,0))
  cycle_start = -1
  while sta and cycle_start == -1:
    v,prev = sta.pop()
    if prev == -1:
      history.pop()
      fin[v] = True
      continue
    if vis[v]: continue
    vis[v] = True
    history.append(v)
    for v2 in G[v]:
      if v2==prev: continue
      if fin[v2]: continue
      if vis[v2]: # 閉路検出
        cycle_start = v2
        break
      sta.append((v2,-1))
      sta.append((v2,v))
  
  cycle = list()
  if cycle_start == -1: return cycle
  while history:
    v = history.pop()
    cycle.append(v)
    if v == cycle_start: break
  return cycle

File: algorithm_library/cpp-from-py/graph/cycle_detect.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> cycle_detect(int start, const vector<vector<int>>& G) {
    int N = (int)G.size() - 1;
    vector<bool> vis(N + 1, false), fin(N + 1, false);
    vector<int> history;
    stack<pair<int, int>> sta;
    sta.emplace(start, 0);
    int cycle_start = -1;
    while (!sta.empty() && cycle_start == -1) {
        auto [v, prev] = sta.top();
        sta.pop();
        if (prev == -1) {
            if (!history.empty()) history.pop_back();
            fin[v] = true;
            continue;
        }
        if (vis[v]) continue;
        vis[v] = true;
        history.push_back(v);
        for (int v2 : G[v]) {
            if (v2 == prev || fin[v2]) continue;
            if (vis[v2]) {
                cycle_start = v2;
                break;
            }
            sta.emplace(v2, -1);
            sta.emplace(v2, v);
        }
    }
    vector<int> cycle;
    if (cycle_start == -1) return cycle;
    while (!history.empty()) {
        int v = history.back();
        history.pop_back();
        cycle.push_back(v);
        if (v == cycle_start) break;
    }
    return cycle;
}
